### DISCUSSION TITLE
GDB crashes when trying to debug

### DISCUSSION THREAD

**Discussion by afternoon** [earlephilhower/arduino-pico#885] [Q&A]
I'm trying to debug my project, but GDB crashes randomly.

- I've tested with the 2.3.3 and 2.5.4 releases of the board package.
- I'm on macOS 12.5.1 on an Intel MacBook Pro.
- I'm using the gdb and openocd versions from the tools installed by the core. (I do have other versions of both on my system, but I'm invoking the right versions with the full path.)
- I'm building using arduino-cli, with a `--build-path` arg pointing to a 'build' subdirectory.
- I'm debugging via a picoprobe.
- The project I'm working on is at https://github.com/afternoon/microgroove in case that's helpful.

GDB seems to crash somewhat randomly. I can start a session and sometimes set breakpoints, but sometimes issuing a command will cause it to crash. E.g. my last session:

```
/Users/ben/Library/Arduino15/packages/rp2040/tools/pqt-gcc/1.4.0-c-0196c06/bin/arm-none-eabi-gdb --eval-command="target extended-remote :3333" build/microgroove.ino.elf
GNU gdb (GDB) 8.2.50.20190202-git
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-apple-darwin14 --target=arm-none-eabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from build/microgroove.ino.elf...
Remote debugging using :3333
0x10008b80 in millis () at /Users/ben/Library/Arduino15/packages/rp2040/hardware/rp2040/2.5.4/cores/rp2040/delay.cpp:58
58	        return to_ms_since_boot(get_absolute_time());
(gdb) b main
Breakpoint 1 at 0x10008e34: file /Users/ben/Library/Arduino15/packages/rp2040/hardware/rp2040/2.5.4//pico-sdk/src/common/pico_stdlib/include/pico/stdlib.h, line 128.
(gdb) info b
Num     Type           Disp Enb Address    What
1       breakpoint     keep y   0x10008e34 in main() at /Users/ben/Library/Arduino15/packages/rp2040/hardware/rp2040/2.5.4//pico-sdk/src/common/pico_stdlib/include/pico/stdlib.h:128
(gdb) d 1
(gdb) b microgroove.ino.cpp:1358
make: *** [debug] Abort trap: 6
```

Any ideas what could be going on here? Or steps I could take to try to root cause the problem? I'm a n00b when it comes to Arduino, the Pico and gdb, so any advice would be much appreciated!

**Comment by earlephilhower** [comment]
Very odd.  I don't have a Mac of any sort but I have heard reports of things working okay.  I'll try to help but in my own Linux (lots and lots) and Window (maybe 3-4hrs total) GDB is running fine.  (Believe me, I've needed while bringing up the core!)

When this crash happens, what happens to the OpenOCD?  In fact, can you manually start each in a separate terminal and check the terminal logs?  I think now you've got a `makefile` which probably starts OpenOCD in the background.

You can also try other versions of GDB.  Anything that speaks ARM should be OK, the ELF format is very standard as is the ARM debug format.

**Reply by afternoon** [reply]
Thanks for the speedy reply! It’s late here in London, so I will try out some stuff tomorrow (other gdb versions, pico-debug, …).

I am running gdb under make, forgot to mention that. Openocd is running in a separate terminal. It seems unphased by the crash, just reports that gdb disconnected. I will post a log tomorrow.

**Reply by afternoon** [reply]
Here's the terminal log from openocd:

```
Open On-Chip Debugger 0.10.0+dev-ge3428fadb-dirty (2021-07-03-09:33)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : only one transport option; autoselect 'swd'
Warn : Transport "swd" was already selected
adapter speed: 5000 kHz

Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : RP2040 Flash Bank Command
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : clock speed 5000 kHz
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x00000001
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x10000001
Info : rp2040.core0: hardware has 4 breakpoints, 2 watchpoints
Info : rp2040.core1: hardware has 4 breakpoints, 2 watchpoints
Info : starting gdb server for rp2040.core0 on 3333
Info : Listening on port 3333 for gdb connections
Info : accepting 'gdb' connection on tcp/3333
Info : RP2040 B0 Flash Probe: 2097152 bytes @10000000, in 512 sectors

Info : New GDB Connection: 1, Target rp2040.core0, state: halted
undefined debug reason 8 - target needs reset
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x00000184 msp: 0x20041f00
Info : dropped 'gdb' connection
```

**Reply by afternoon** [reply]
Using homebrew's gdb (10.3-2021.07) works great. I can now debug the things!

Something about the version of gdb and my system are getting stuck. It might be worth expanding the debugging section in the docs a little. I'll send a PR if I get a moment.

Thanks again for your help, and thank you so much for arduino-pico. It's a fantastic tool to build on and its clear a ton of work went into it. 😍
